File: apps/node/credentials.go
package main

// Per-device credentials (#1042 phase 2).
//
// The daemon used to authenticate every device with one shared bearer
// token. It now mints each device its own credential — installers already
// embed one — and a node still running on the shared token trades it in
// band: the /devices/register reply (the 60s heartbeat) carries
// `credential: {action: "upgrade"}`, the node POSTs /auth/upgrade, persists
// the returned `tdc1.…` token to its config (next to the pinned
// fingerprint) and uses it from then on. The same exchange rotates a
// per-device credential when the operator asks (`action: "rotate"`).
//
// The swap is persist-then-adopt: a token that could not be written to the
// config is never used, so a restart can't strand the node on a credential
// it no longer has. The daemon keeps the old credential valid until the new
// one is first presented, so a lost reply is harmless too.

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"regexp"
	"strings"
)

// deviceCredentialRE is the per-device credential shape (protocol/fixtures/
// auth_v1.json credentialPattern). Anything else is the shared token.
var deviceCredentialRE = regexp.MustCompile(`^tdc1\.[0-9a-f]{16}\.[A-Za-z0-9_-]{43}$`)

func isDeviceCredential(token string) bool {
	return deviceCredentialRE.MatchString(token)
}

// registerReply is the POST /devices/register response.
type registerReply struct {
	OK         bool            `json:"ok"`
	DeviceID   string          `json:"deviceId"`
	Credential *credentialHint `json:"credential,omitempty"`
}

type credentialHint struct {
	Action string `json:"action"`
}

// credentialAction is what the node should do about the reply's hint given
// the token it is using: "upgrade", "rotate", or "" (nothing — including
// for any action this build does not know, which must be ignored).
func (r registerReply) credentialAction(token string) string {
	if r.Credential == nil {
		return ""
	}
	switch r.Credential.Action {
	case "upgrade":
		if !isDeviceCredential(token) {
			return "upgrade"
		}
	case "rotate":
		if isDeviceCredential(token) {
			return "rotate"
		}
	}
	return ""
}

// upgradeRequestBody is the POST /auth/upgrade payload. A node only ever
// asks for the device scope.
func upgradeRequestBody(deviceID string) map[string]any {
	return map[string]any{
		"deviceId": deviceID,
		"client":   "node",
		"scopes":   []string{"device"},
	}
}

type upgradeReply struct {
	OK           bool     `json:"ok"`
	Token        string   `json:"token"`
	CredentialID string   `json:"credentialId"`
	DeviceID     string   `json:"deviceId"`
	Scopes       []string `json:"scopes"`
	Error        string   `json:"error"`
}

// parseUpgradeReply decodes and validates an /auth/upgrade reply: it must
// carry a well-formed credential bound to THIS node.
func parseUpgradeReply(raw []byte, deviceID string) (upgradeReply, error) {
	var r upgradeReply
	if err := json.Unmarshal(raw, &r); err != nil {
		return r, fmt.Errorf("upgrade reply: %w", err)
	}
	if !r.OK {
		return r, fmt.Errorf("upgrade refused: %s", r.Error)
	}
	if !isDeviceCredential(r.Token) {
		return r, errors.New("upgrade reply carried no valid credential")
	}
	if r.DeviceID != deviceID {
		return r, fmt.Errorf("credential is for %q, not this node (%q)", r.DeviceID, deviceID)
	}
	return r, nil
}

// token is the bearer in use right now (it changes on upgrade/rotation).
func (n *Node) token() string {
	n.tokenMu.RLock()
	defer n.tokenMu.RUnlock()
	return n.cfg.Token
}

// saveConfig writes the config under the token lock, so a save never
// serializes a half-swapped credential.
func (n *Node) saveConfig() error {
	n.tokenMu.Lock()
	defer n.tokenMu.Unlock()
	return n.cfg.Save()
}

// adoptCredential persists `token` and only then starts using it.
func (n *Node) adoptCredential(token string) error {
	n.tokenMu.Lock()
	defer n.tokenMu.Unlock()
	prev := n.cfg.Token
	n.cfg.Token = token
	if err := n.cfg.Save(); err != nil {
		n.cfg.Token = prev
		return err
	}
	return nil
}

// maybeUpgradeCredential acts on a register reply's credential hint. Runs
// at most one exchange at a time; a daemon without per-device credentials
// (404) or a config the node cannot write turns it off for this run.
func (n *Node) maybeUpgradeCredential(ctx context.Context, reply registerReply) {
	action := reply.credentialAction(n.token())
	if action == "" || n.upgradeDisabled.Load() {
		return
	}
	if !n.upgrading.CompareAndSwap(false, true) {
		return
	}
	defer n.upgrading.Store(false)

	var raw json.RawMessage
	err := n.postJSON(ctx, "/auth/upgrade", upgradeRequestBody(n.DeviceID), &raw)
	if err != nil {
		log.Printf("credential %s failed: %v", action, err)
		if strings.Contains(err.Error(), "HTTP 404") {
			n.upgradeDisabled.Store(true)
		}
		return
	}
	up, err := parseUpgradeReply(raw, n.DeviceID)
	if err != nil {
		log.Printf("credential %s failed: %v", action, err)
		return
	}
	if err := n.adoptCredential(up.Token); err != nil {
		// Never use a credential the config does not hold: stay on the old
		// one (still valid daemon-side) and stop asking for this run.
		log.Printf("credential %s: could not persist the new credential (%v) — keeping the current one", action, err)
		n.upgradeDisabled.Store(true)
		return
	}
	log.Printf(
		"credential %s: now using per-device credential %s (scopes %v)",
		action, up.CredentialID, up.Scopes,
	)
}

// credentialKind describes the configured bearer for `talon-node status`.
func credentialKind(token string) string {
	switch {
	case token == "":
		return "(none)"
	case isDeviceCredential(token):
		return "per-device credential " + strings.SplitN(token, ".", 3)[1]
	default:
		return "shared bridge token (legacy — upgraded automatically on connect)"
	}
}

File: apps/companion/lib/src/models/credentials.dart
/// Per-device bridge credentials (#1042 phase 2).
///
/// The daemon used to authenticate every client with one shared token. It
/// now mints each device its own credential — `tdc1.<id>.<secret>` — bound
/// to that device's mesh id and scoped (`device`, `client`, `operator`).
/// Pairing links already carry one; a profile still holding the shared
/// token trades it in band (`POST /auth/upgrade`) the first time it
/// connects, and stores the result in its connection profile in place of
/// the shared token. The wire shapes live in protocol/fixtures/auth_v1.json.
library;

final RegExp _deviceCredential = RegExp(
  r'^tdc1\.[0-9a-f]{16}\.[A-Za-z0-9_-]{43}$',
);

/// True for a per-device credential, false for the shared token (or none).
bool isDeviceCredential(String? token) =>
    token != null && _deviceCredential.hasMatch(token);

/// `GET /auth/whoami`: which credential this connection uses and whether the
/// daemon wants it upgraded (shared token) or rotated.
class CredentialStatus {
  final String kind;
  final List<String> scopes;
  final String? credentialId;
  final String? deviceId;

  /// `upgrade`, `rotate`, or null. Unknown actions come through verbatim and
  /// must be ignored by callers — see [CredentialStatus.wantsNewCredential].
  final String? action;

  const CredentialStatus({
    required this.kind,
    required this.scopes,
    this.credentialId,
    this.deviceId,
    this.action,
  });

  factory CredentialStatus.fromJson(Map<String, dynamic> j) => CredentialStatus(
        kind: j['kind'] is String ? j['kind'] as String : 'unknown',
        scopes: j['scopes'] is List
            ? (j['scopes'] as List).whereType<String>().toList()
            : const [],
        credentialId: j['credentialId'] as String?,
        deviceId: j['deviceId'] as String?,
        action: j['action'] as String?,
      );

  /// Whether a client holding [token] should ask for a new credential now:
  /// an upgrade off the shared token, or a rotation of its own credential.
  bool wantsNewCredential(String? token) {
    if (action == 'upgrade') return !isDeviceCredential(token);
    if (action == 'rotate') return isDeviceCredential(token);
    return false;
  }
}

/// A `POST /auth/upgrade` reply: the new credential. The reply is the only
/// copy of [token] anywhere — the daemon keeps a hash.
class CredentialGrant {
  final String token;
  final String credentialId;
  final String deviceId;
  final List<String> scopes;

  const CredentialGrant({
    required this.token,
    required this.credentialId,
    required this.deviceId,
    required this.scopes,
  });

  /// Throws [FormatException] for a refusal or a reply without a well-formed
  /// credential, so a bad reply can never replace a working token.
  factory CredentialGrant.fromJson(Map<String, dynamic> j) {
    final token = j['token'];
    if (j['ok'] != true || token is! String || !isDeviceCredential(token)) {
      throw FormatException(
        'No credential in upgrade reply: ${j['error'] ?? 'malformed'}',
      );
    }
    return CredentialGrant(
      token: token,
      credentialId: (j['credentialId'] ?? '') as String,
      deviceId: (j['deviceId'] ?? '') as String,
      scopes: j['scopes'] is List
          ? (j['scopes'] as List).whereType<String>().toList()
          : const [],
    );
  }
}
